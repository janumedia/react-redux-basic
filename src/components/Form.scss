form {
    background-color: #ECECEC;
    padding: 10px;
    margin: 0 auto;
    label, input {
        margin: 10px 0;
    }
    input, button {
        padding: 10px;
    }
    input {
        width: 100%;
    }
    button {
        text-transform: uppercase;
    }
}