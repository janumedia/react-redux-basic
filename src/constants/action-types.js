const asyncActionTypes = type => ({
    PENDING: `${type}_PENDING`,
    SUCCESS: `${type}_SUCCESS`,
    FAILED: `${type}_FAILED`
});

export const GET_USERS = asyncActionTypes('GET_USERS');
export const ADD_USER = asyncActionTypes('ADD_USER');
export const DELETE_USER = 'DELETE_USER';