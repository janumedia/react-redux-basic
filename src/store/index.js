import { createStore, applyMiddleware } from 'redux';

import userReducer from '../reducers/userReducer';
import userMiddleWare from '../middleware/userMiddleWare';

export default createStore (userReducer, applyMiddleware(userMiddleWare));