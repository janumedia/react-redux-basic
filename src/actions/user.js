import { ADD_USER, DELETE_USER } from "../constants/action-type";

export const addUser = (newUser) => ({
    type: ADD_USER,
    payload: newUser
})

export const deleteUser = (userID) => ({
    type: DELETE_USER,
    payload: userID
})
