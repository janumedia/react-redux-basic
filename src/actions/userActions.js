import * as types from "../constants/action-types";

export const addUser = (newUser) => ({
    type: types.ADD_USER,
    payload: newUser
})

export const deleteUser = (userID) => ({
    type: types.DELETE_USER,
    payload: userID
})

export const getUsers = () => ({
    type: types.GET_USERS,
    payload: {url: '/data/users.json'}
})
