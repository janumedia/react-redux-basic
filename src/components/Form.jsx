import React from 'react';
import { connect } from 'react-redux';

import { addUser } from '../actions/userActions';
import { ADD_USER } from '../constants/action-types';
import './Form.scss';

const mapDispatchToProps = dispatch => {
    return {
        addUser: user => dispatch(addUser(user))
    };
};

class Form extends React.Component {
    
    constructor(props) {
        super(props);

        this.state = {
            name: '',
            role: ''
        }
    }
    
    handleInputChange(e) {
        this.setState(
            { [e.target.id]: e.target.value }
        )
    }

    handleFormSubmit(e) {
        e.preventDefault();
        const {name, role} = this.state;
        this.props.addUser({name, role}).then(status => {
            this.setState({status: status});
            if(status === ADD_USER.SUCCESS) {
                this.setState({
                    name: '',
                    role: ''
                })
            }
        })
    }
    
    render() {
        return (
            <div className="Form">
                <form onSubmit={this.handleFormSubmit.bind(this)}>
                    <div>
                        <label>Name</label>
                        <input type="text" id="name" value={this.state.name} onChange={this.handleInputChange.bind(this)} />
                        <label>Role</label>
                        <input type="text" id="role" value={this.state.role} onChange={this.handleInputChange.bind(this)} />
                    </div>
                    <button type="submit">Add User</button>
                </form>
                {this.props.children}
            </div>
        );
    }
};

export default connect(null, mapDispatchToProps)(Form);