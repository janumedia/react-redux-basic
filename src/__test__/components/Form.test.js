import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

import store from '../../store';
import * as userActions from '../../actions/userActions';
import { ADD_USER } from '../../constants/action-types';
import Form from '../../components/Form';

jest.useFakeTimers();

describe('Form component', () => {

    let wrapper, form, inputName, inputRole, addUserSpy;

    beforeEach(() => {
        wrapper = shallow(<Form store={store}/>).dive();
        form  = wrapper.find('form');
        inputName = wrapper.find('#name');
        inputRole = wrapper.find('#role');
        addUserSpy = jest.spyOn(userActions, 'addUser');
    });

    afterEach(() => {
        addUserSpy.mockRestore();
        //jest.resetAllMocks();
        jest.clearAllTimers();
    })

    it('should render properly', () => {
        expect(wrapper.type()).toEqual('div');
        expect(wrapper.hasClass('Form')).toBe(true);
        expect(wrapper.find('#name')).toHaveLength(1);
        expect(wrapper.find('#role')).toHaveLength(1);
        expect(wrapper.find('button')).toHaveLength(1);
        expect(toJson(wrapper)).toMatchSnapshot();
    });
    
    it('should trigger "addUser', () => {
        form.simulate('submit', {preventDefault:() => {}});
        expect(addUserSpy).toHaveBeenCalledTimes(1);
    })

    it('should return "ADD_USER.FAILED" if NOT fill both "name" and "role"', async() => {
        form.simulate('submit', {preventDefault:() => {}});
        
        jest.runOnlyPendingTimers();
        expect(addUserSpy).toHaveBeenCalledTimes(1);
        
        //simulate failed mock resolved
        const mockResolveAddUser = addUserSpy.mockResolvedValueOnce();
        await mockResolveAddUser();

        expect(wrapper.state('status')).toEqual(ADD_USER.FAILED);
    })

    
    it('should reset form if return "ADD_USER.SUCCES"', async () => {
        
        inputName.simulate('change', {target:{id:"name", value:"Dave"}});
        inputRole.simulate('change', {target:{id:"role", value:"Engineer"}});
        
        form.simulate('submit', {target:form, preventDefault:() => {}});
        jest.runOnlyPendingTimers();

        const mockResolveAddUser = addUserSpy.mockResolvedValueOnce();
        await mockResolveAddUser();
        
        if(wrapper.state('status') === ADD_USER.SUCCESS)
        {
            expect(wrapper.state('name')).toEqual('');
            expect(wrapper.state('role')).toEqual('');
        } else {
            expect(wrapper.state('name')).toEqual('Dave');
            expect(wrapper.state('role')).toEqual('Engineer');
        }
    });

    it('should return "ADD_USER.FAILED if "name" already exist', async () => {
        
        inputName.simulate('change', {target:{id:"name", value:"Dave"}});
        inputRole.simulate('change', {target:{id:"role", value:"Lead Engineer"}});
        
        form.simulate('submit', {target:form, preventDefault:() => {}});
        jest.runOnlyPendingTimers();

        const mockResolveAddUser = addUserSpy.mockResolvedValueOnce();
        await mockResolveAddUser();

        expect(wrapper.state('status')).toEqual(ADD_USER.FAILED);
    });
})