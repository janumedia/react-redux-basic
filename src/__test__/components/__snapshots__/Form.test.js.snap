// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form component should render properly 1`] = `
<div
  className="Form"
>
  <form
    onSubmit={[Function]}
  >
    <div>
      <label>
        Name
      </label>
      <input
        id="name"
        onChange={[Function]}
        type="text"
        value=""
      />
      <label>
        Role
      </label>
      <input
        id="role"
        onChange={[Function]}
        type="text"
        value=""
      />
    </div>
    <button
      type="submit"
    >
      Add User
    </button>
  </form>
</div>
`;
